velox/functions/sparksql/SIMDJsonFunctions.h

```diff
@@ -0,0 +1,195 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/prestosql/SIMDJsonFunctions.h"
+
+using namespace simdjson;
+
+namespace facebook::velox::functions::sparksql {
+
+template <typename T>
+struct SIMDGetJsonObjectFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const std::vector<TypePtr>& /*inputTypes*/,
+      const core::QueryConfig& config,
+      const arg_type<Varchar>* /*json*/,
+      const arg_type<Varchar>* jsonPath) {
+    if (jsonPath != nullptr) {
+      if (jsonPath->size() > 1 && jsonPath->data()[0] == '$') {
+        jsonPath_ = removeSingleQuotes(
+            std::string_view(jsonPath->data() + 1, jsonPath->size() - 1));
+      }
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& json,
+      const arg_type<Varchar>& jsonPath) {
+    // Spark requires the first char in jsonPath is '$'.
+    if (jsonPath.size() < 1 || jsonPath.data()[0] != '$') {
+      return false;
+    }
+    // jsonPath is "$".
+    if (jsonPath.size() == 1) {
+      result.append(json);
+      return true;
+    }
+    ParserContext ctx(json.data(), json.size());
+    ctx.parseDocument();
+    auto rawResult = jsonPath_.has_value()
+        ? ctx.jsonDoc.at_path(jsonPath_.value().data())
+        : ctx.jsonDoc.at_path(
+              removeSingleQuotes(
+                  std::string_view(jsonPath.data() + 1, jsonPath.size() - 1))
+                  .data());
+    if (rawResult.error()) {
+      return false;
+    }
+    auto error = extractStringResult(rawResult, result);
+    if (error) {
+      return false;
+    }
+
+    const char* currentPos;
+    ctx.jsonDoc.current_location().get(currentPos);
+    return isValidEndingCharacter(currentPos);
+  }
+
+ private:
+  // Spark's json path requires field name surrounded by single quotes if it is
+  // specified in "[]". But simdjson lib requires not. This method just removes
+  // such single quotes, e.g., converts "['a']['b']" to "[a][b]".
+  FOLLY_ALWAYS_INLINE std::string removeSingleQuotes(
+      std::string_view jsonPath) {
+    std::string result(jsonPath.data(), jsonPath.size());
+    size_t pairEnd = 0;
+    while (true) {
+      auto pairBegin = result.find("['", pairEnd);
+      if (pairBegin == std::string::npos) {
+        break;
+      }
+      pairEnd = result.find("]", pairBegin);
+      if (pairEnd == std::string::npos || result[pairEnd - 1] != '\'') {
+        return "-1";
```

Comment:
This seems a bit hacky. Are you relying on simdjson to reject this path later? It would be cleaner to raise an exception here and provide a clear error message.

This function seems to be working around a limitation on simdjson. Have you already asked in simdjson project if that limitation can be removed?  Anyway, please, update PR description to explain this limitation and how you are working around it.

Reply:
> Are you relying on simdjson to reject this path later? It would be cleaner to raise an exception here and provide a clear error message.

@mbasmanova, yes, we depend on simdjson to return error code for illegal path, then this function will return NULL result at last, instead of throwing exception.

> Have you already asked in simdjson project if that limitation can be removed?

I just left a comment in simdjson community to discuss this limitation:
https://github.com/simdjson/simdjson/issues/2070#issuecomment-2079688880

Reply:
@PHILO-HE 

The API that returns NULL when json or path are invalid is challenging. Throwing an error would be better.

Note that simdjson has limited support for JSONPath, hence, it rejects both invalid and valid-but-not-supported paths. If Spark supports full JSONPath spec or a wider subset of the spec than simdjson, then this implementation will produce incorrect results.

Hence, some questions.

- What subset of JSONPath is supported in Spark?
- Is there any particular reason to not use SIMDJsonExtract[Scalar]Function directly or implement this in a similar way by leveraging velox/functions/prestosql/json/SIMDJsonExtractor.h ?
- Does Spark validate JSON document fully and return null if it is not valid? simdjson doesn't do that and may succeed at extracting the value from an invalid document. This will then cause Gluten to produce wrong results.

Reply:
> @PHILO-HE
> 
> The API that returns NULL when json or path are invalid is challenging. Throwing an error would be better.
> 
> Note that simdjson has limited support for JSONPath, hence, it rejects both invalid and valid-but-not-supported paths. If Spark supports full JSONPath spec or a wider subset of the spec than simdjson, then this implementation will produce incorrect results.
> 
> Hence, some questions.
> 
> * What subset of JSONPath is supported in Spark?

@mbasmanova, thanks for your comment!
Spark has its own code to parse Json path. I have addressed all inconsistency found in Spark tests. For example, ['name']['id'] is supported by Spark, but not by Simdjson. This pr fixed this inconsistency by just pre-processing the path. So I feel it's better to return NULL instead of throwing exception for invalid path.

> * Is there any particular reason to not use SIMDJsonExtract[Scalar]Function directly or implement this in a similar way by leveraging velox/functions/prestosql/json/SIMDJsonExtractor.h ?

The proposed pr has some handlings to align with Spark, e.g., check the validity of JSON document. And with this proposed patch, all Spark test cases can pass.

> * Does Spark validate JSON document fully and return null if it is not valid? simdjson doesn't do that and may succeed at extracting the value from an invalid document. This will then cause Gluten to produce wrong results.

Yes, Spark fully validates JSON document. Simdjson only checks the validity before the position of extracted result. That's why this pr checked the characters after this position. Though it's not a full validation, it can meet our Spark users' requirement.